fem/particleset.hpp

```diff
@@ -0,0 +1,521 @@
+   void SetTagRef(int t, int *tag_data);
+
+   /// Particle equality operator.
+   bool operator==(const Particle &rhs) const;
+
+   /// Particle inequality operator.
+   bool operator!=(const Particle &rhs) const { return !operator==(rhs); }
+
+   /// Print all particle data to \p os.
+   void Print(std::ostream &os=mfem::out) const;
+};
+
+/** @brief ParticleSet initializes and manages data associated with particles.
+ *
+ *  @details Particles are inherently initialized to have a position and an ID,
+ *  and optionally can have any number of Vector (of arbitrary vdim) and scalar
+ *  integer data in the form of @b fields and @b tags respectively. All particle
+ *  data are internally stored in a Struct-of-Arrays fashion, as elaborated on below.
+ *
+ *  @par Coordinates:
+ *  All particle coordinates are stored in a \ref ParticleVector with vector
+ *  dimension equal to the spatial dimension, ordered either byNODES or byVDIM.
+ *
+ *  @par IDs:
+ *  Each particle is assigned a unique global ID of type unsigned int. In
+ *  parallel, IDs are initialized starting with @b rank and striding by @b size.
+ *
+ *  @par Fields:
+ *  Fields represent scalar or vector \ref real_t data to be associated with
+ *  each particles, such as mass, momentum, or moment. For a given field, all
+ *  particle data is stored in a single \ref ParticleVector with a given
+ *  vector dimension (1 for scalar data) and \ref Ordering::Type (byNODES or
+ *  byVDIM).
+ *
+ *  @par Tags:
+ *  Tags represent an integer to be associated with each particle. For a given
+ *  tag, all particle integers are stored in a single \ref Array<int>.
+ *
+ *  @par Names:
+ *  Each field and tag can optionally be given a name (string) to be used when
+ *  printing particle data in CSV format using \ref PrintCSV.
+ *
+ */
+class ParticleSet
+{
+private:
+   /// Constructs an Array of size N filled with Ordering::Type o.
+   static Array<Ordering::Type> GetOrderingArray(Ordering::Type o, int N);
+   /// Returns default field name for field index i. "Field_{i}"
+   static std::string GetDefaultFieldName(int i);
+   /// Returns default tag name for tag index i. "Tag_{i}"
+   static std::string GetDefaultTagName(int i);
+   /// Constructs an Array of size N filled with nullptr.
+   static Array<const char*> GetEmptyNameArray(int N);
+#ifdef MFEM_USE_MPI
+   static unsigned int GetRank(MPI_Comm comm_);
+   static unsigned int GetSize(MPI_Comm comm_);
+#endif // MFEM_USE_MPI
+
+protected:
+
+   /// Stride for IDs (when new particles are added).
+   /// In parallel, this defaults to the number of MPI ranks.
+   const unsigned int id_stride;
+
+   /// Current ID to be assigned to the next particle added.
+   /// In parallel, this starts locally as the rank and increments with
+   /// id_stride, ensuring a global unique identifier whenever a particle is
+   /// added.
+   unsigned int id_counter;
+
+   /// Array holding all particles' globally unique ID.
+   Array<unsigned int> ids;
```

Comment:
`unsigned int` will easily not be enough for global IDs, maybe we should consider a longer type. Potentially it could be optional like Hypre BigInt 🤔 .